### PR TITLE
[inductor] Fix scalar-broadcast scatter zero-mask analysis

### LINKED ISSUE #178871: Incorrect output for non-power-of-2-sized fusion of scalar broadcast and scatter-add
### 🐛 Describe the bug

When inductor fuses a scalar broadcast with `scatter_reduce(..., reduce="sum")`, the generated Triton kernel uses `mask=None` on `tl.atomic_add` instead of `mask=xmask`. `XBLOCK` is always a power of 2, so when `N` is not a power of 2 there are `XBLOCK - N` out-of-bounds threads — they each add the broadcast scalar to the output, giving a wrong result with no error raised.

The error is exactly `(next_power_of_2(N) - N) * scalar` and is zero whenever `N` is a power of 2.

**Minimal reproducer:**

```python
import torch


def f(x, bias, idx):
    src = x + bias  # scalar broadcast fused with scatter by inductor
    out = torch.zeros(1, 1, device=x.device, dtype=x.dtype)
    return out.scatter_reduce(
        0, idx.unsqueeze(-1).expand_as(src), src, "sum", include_self=True
    )


for N in range(1, 65):
    device = torch.device("cuda")

    x    = torch.randn(N, 1, device=device)
    bias = torch.tensor([0.020368], device=device)
    idx  = torch.zeros(N, dtype=torch.long, device=device)

    ref = f(x, bias, idx)
    torch.compiler.reset()
    test = torch.compile(f, backend="inductor", fullgraph=True, dynamic=False)(x, bias, idx)

    diff      = test.item() - ref.item()
    next_pow2 = 1 << (N - 1).bit_length()
    expected  = (next_pow2 - N) * bias.item()
    print(f"N={N:2d}  diff={diff:+.6f}  expected={expected:+.6f}")
```

The generated kernel (trimmed):

```python
tmp0 = tl.load(in_ptr0 + x0, xmask)  # x     — masked, 0.0 for OOB threads
tmp1 = tl.load(in_ptr1 + 0)           # bias  — scalar, no mask needed
tmp2 = tmp0 + tmp1                    # bias for OOB threads (tmp0 == 0.0)

tl.atomic_add(out_ptr0 + ..., tmp2, None,  sem='relaxed')  # BUG: should be xmask
```

OOB threads add `bias` unconditionally. Total error: `(XBLOCK - N) * bias`.

### Error logs

No exception — the output is silently wrong.

```
N= 1  diff=+0.000000  expected=+0.000000
N= 2  diff=+0.000000  expected=+0.000000
N= 3  diff=+0.020368  expected=+0.020368
N= 4  diff=+0.000000  expected=+0.000000
N= 5  diff=+0.061104  expected=+0.061104
N= 6  diff=+0.040736  expected=+0.040736
N= 7  diff=+0.020368  expected=+0.020368
N= 8  diff=+0.000000  expected=+0.000000
N= 9  diff=+0.142576  expected=+0.142576
...
N=16  diff=+0.000000  expected=+0.000000
N=17  diff=+0.305520  expected=+0.305520
...
N=32  diff=+0.000000  expected=+0.000000
N=33  diff=+0.631407  expected=+0.631408
...
N=48  diff=+0.325887  expected=+0.325888
...
N=64  diff=+0.000000  expected=+0.000000
```

### Versions

```
PyTorch version: 2.10.0+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 12.3.0-1ubuntu1~22.04.2) 12.3.0
Python version: 3.11.15 | packaged by conda-forge | (main, Mar  5 2026, 16:45:40) [GCC 14.3.0] (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 12.8.93
GPU models and configuration: GPU 0: NVIDIA TITAN V
Nvidia driver version: 570.195.03
triton: 3.6.0
```

cc @ezyang @gchanan @kadeng @msaroufim @pragupta @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by keon**
## Summary
Fixes a silent correctness issue in the zero-mask analysis used for prologue fusion around scatter-reduce kernels.

`PreservesZeros.load()` was assuming that all prologue loads behave like masked tensor loads and therefore produce `0` for out-of-bounds lanes. That is not true for scalar / numel-1 broadcast inputs such as the fused `bias` in the issue reproducer.

When that scalar load was treated as zero-preserving, Inductor could incorrectly conclude that the fused prologue preserved masked-out values, which in turn could drop the effective mask on the scatter atomic-add path for non-power-of-2 launch shapes.

This change makes the analysis conservative for known `numel == 1` loads by treating them as unknown values instead of masked zeros.

Fixes #178871

## Test Plan
- Standalone reproducer for the zero-mask analysis failed before the patch and passed after it
- Added regression coverage in `test/inductor/test_scatter_optimization.py`
- `python -m py_compile torch/_inductor/analyze_preserves_zero_mask.py test/inductor/test_scatter_optimization.py`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Review by eellison — CHANGES_REQUESTED** [review]
Are you sure this is relevant ? that pass should only be for template prologue fusion.
